common/persistence/visibility_hybrid_manager.go

```diff
@@ -0,0 +1,554 @@
+	}
+	return manager.ListWorkflowExecutions(ctx, request)
+}
+
+func (v *visibilityHybridManager) ScanWorkflowExecutions(
+	ctx context.Context,
+	request *ListWorkflowExecutionsByQueryRequest,
+) (*ListWorkflowExecutionsResponse, error) {
+	override := ctx.Value(ContextKey)
+	v.logUserQueryParameters(userParameters{
+		operation:    string(Operation.LIST),
+		domainName:   request.Domain,
+		closeStatus:  6, // 6 means not set closeStatus.
+		customQuery:  request.Query,
+		earliestTime: -1,
+		latestTime:   -1,
+	}, request.Domain, override != nil)
+
+	manager, shadowMgr := v.chooseVisibilityManagerForRead(ctx, request.Domain)
+	if shadowMgr != nil {
+		go shadow(shadowMgr.ScanWorkflowExecutions, request, v.logger)
+	}
+	return manager.ScanWorkflowExecutions(ctx, request)
+}
+
+func (v *visibilityHybridManager) CountWorkflowExecutions(
+	ctx context.Context,
+	request *CountWorkflowExecutionsRequest,
+) (*CountWorkflowExecutionsResponse, error) {
+	override := ctx.Value(ContextKey)
+	v.logUserQueryParameters(userParameters{
+		operation:    string(Operation.COUNT),
+		domainName:   request.Domain,
+		closeStatus:  6, // 6 means not set closeStatus.
+		customQuery:  request.Query,
+		earliestTime: -1,
+		latestTime:   -1,
+	}, request.Domain, override != nil)
+
+	manager, shadowMgr := v.chooseVisibilityManagerForRead(ctx, request.Domain)
+	if shadowMgr != nil {
+		go shadow(shadowMgr.CountWorkflowExecutions, request, v.logger)
+	}
+	return manager.CountWorkflowExecutions(ctx, request)
+}
+
+func (v *visibilityHybridManager) chooseVisibilityManagerForRead(ctx context.Context, domain string) (VisibilityManager, VisibilityManager) {
+	if visOverride, ok := ctx.Value(ContextKey).(string); ok && visOverride != "" && visOverride != "Default" {
+		v.logger.Info(fmt.Sprintf("Visibility Migration log: %s visibility manager was chosen for read.", visOverride))
+		if v.visibilityMgrs[visOverride] != nil {
+			return v.visibilityMgrs[visOverride], nil
+		} // if the override is not valid, fall back to default
+	}
+	var visibilityMgr, shadowMgr VisibilityManager
+	stores := strings.Split(v.readVisibilityStoreName(domain), ",")
+	readStore := stores[0] // if read stores have more than 1, the others will go shadow read
+	if v.visibilityMgrs[readStore] != nil {
+		visibilityMgr = v.visibilityMgrs[readStore]
+	} else {
+		v.logger.Warn("domain is configured to read from advanced visibility but it's not available, fall back to basic visibility",
+			tag.WorkflowDomainName(domain))
+		visibilityMgr = v.visibilityMgrs[dbVisStoreName] //db will always be available
+	}
+
+	if len(stores) > 1 {
+		shadowMgr = v.visibilityMgrs[stores[1]]
+	}
+
+	return visibilityMgr, shadowMgr
+}
+
+func shadow[ReqT any, ResT any](f func(ctx context.Context, request ReqT) (ResT, error), request ReqT, logger log.Logger) {
+	ctxNew, cancel := context.WithTimeout(context.Background(), 2*time.Minute) // don't want f to run too long
```

Comment:
2 minutes sounds a lot here. we create a shadow goroutine for each request and letting them hang for 2 minutes will accumulate if shadow destination is slow. Can we limit the shadow goroutine deadline the same as original request's deadline (based on given `ctx`)?

Reply:
I talked with Bowen, this is because Pinot used to have high latency. I will update this

Reply:
Reduced to 30s when we test the performance, will change it when having more metrics